### PR TITLE
Prebuild with z3

### PR DESCRIPTION
**Pull request by rlisahuang**
work in progress

TODOs:
- [x] connect file i/o with the remark plugin
- [x] make z3 work -- probably not using `eval_smtlib2_string`

### CODE REVIEW

**Comment by NikolajBjorner** [comment]
At some point I presume I will be testing this locally as well using the dev environment by toggling the z3 functionality on / off?

**Comment by pelikhan** [comment]
Running a build will run z3 on every z3 snippet
________________________________

**Comment by rlisahuang** [comment]
@pelikhan Our original logic without timeouts does not work because there is one snippet that takes forever to run. So I tried to add a 500ms timeout (ideally should be 3-5 minutes, keeping it short just to see if it works) on long evals. It is buggy at the moment for the following reasons:
- The evals are still running asynchronously despite our changes. I only discovered this by putting breakpoints inside the `try/catch` block.
- The timeout error, thrown in the `try` block, cannot be caught by the following `catch` block.

File: website/src/remark/render-code-blocks.mjs [hunk 1]
```diff
@@ -13,15 +13,21 @@ import { execSync } from 'child_process';
 
 
 async function getOutput(input) {
+    const timeOut = 30000;
     const hashObj = createHash('sha1');
 
-    const hash = hashObj.update(input + z3pkg.version).digest('hex');
+    const hash = hashObj.update(input).update(z3pkg.version).update(timeOut);
```
Comment (pelikhan) [inline website/src/remark/render-code-blocks.mjs:19]: add timeout to hash

File: website/src/remark/render-code-blocks.mjs [hunk 2]
```diff
@@ -13,15 +13,21 @@ import { execSync } from 'child_process';
...
+    
+    hash.digest('hex');
+
+    ensureDirSync(`./build/solutions/${hash}`);
```
Comment (pelikhan) [inline website/src/remark/render-code-blocks.mjs:23]: move `./build/solutions/${hash}` to local variable

File: website/src/remark/render-code-blocks.mjs [hunk 3]
```diff
@@ -34,20 +40,24 @@ async function getOutput(input) {
...
-    const inputObj = JSON.stringify({ arg: input });
+    const inputObj = { input: input };
+    writeJsonSync(pathIn, inputObj);
 
-    const command = `node ./src/remark/run-z3.js ${inputObj}`;
+    const command = `node ./src/remark/run-z3.js ${pathIn}`;
 
     try {
-        output = execSync(command, { timeout: 5000 }).toString();
+        output = execSync(command, { timeout: 30000 });
```
Comment (pelikhan) [inline website/src/remark/render-code-blocks.mjs:51]: use timeout variable

**Comment by NikolajBjorner** [comment]
> either z3 wasm needs to support timeout or you need to run it as a separate child proces

Z3 itself has:

- Z3_interrupt - you can call from your separate thread.
- a timeout parameter you can set on the solver object

**Comment by pelikhan** [comment]
we went down the child_process road, which also allows us to run concurrently.

**Comment by pelikhan** [comment]
Always render output if present imo. We can improve this later.

**Comment by rlisahuang** [comment]
It took 18min to build without caching in github actions / concurrency. Ready to merge.

**Comment by pelikhan** [comment]
Next or: add caching in action!

**Pull request merged**